File: flutter-src/lib/services/external_gtfs/transit_data_provider.dart
// TransitDataProvider — miroir Dart de lib/external-gtfs/transit-data-provider.js.
// Couche unique de résolution des horaires. Ordre de priorité :
//   1. CETUD CURRENT (currentOfficial, autorité CETUD)
//   2. autre source officielle actuelle
//   3. PassBi CURRENT si réellement prouvé (currentApplication)
//   4. open data actuelle
//   5. estimation par fréquence documentée (ESTIMATED)
//   6. UNKNOWN
// Les sources HISTORICAL ne participent JAMAIS à une réponse « maintenant ».

import 'feed_provenance.dart';
import 'gtfs_feed.dart';
import 'gtfs_schedule_service.dart';
import 'gtfs_time.dart';

class ProviderRoles {
  static const String currentOfficial = 'currentOfficial';
  static const String currentApplication = 'currentApplication';
  static const String currentOpenData = 'currentOpenData';
  static const String currentFrequency = 'currentFrequency';
  static const String historicalReference = 'historicalReference';
  static const List<String> all = <String>[
    currentOfficial,
    currentApplication,
    currentOpenData,
    currentFrequency,
    historicalReference,
  ];
  static const List<String> current = <String>[
    currentOfficial,
    currentApplication,
    currentOpenData,
    currentFrequency,
  ];
  static const Map<String, String> requiredSourceType = <String, String>{
    currentOfficial: SourceTypes.institutional,
    currentApplication: SourceTypes.application,
    currentOpenData: SourceTypes.openData,
  };
  static const Map<String, String> level = <String, String>{
    currentOfficial: ProvenanceLevels.officialStaticCurrent,
    currentApplication: ProvenanceLevels.sourceApplicationCurrent,
    currentOpenData: ProvenanceLevels.openDataCurrent,
    currentFrequency: ProvenanceLevels.estimated,
    historicalReference: ProvenanceLevels.historicalReference,
  };
}

/// Normalisation de comparaison (accents, casse, ponctuation).
String normalizeText(String? s) {
  if (s == null) return '';
  const String from = 'àáâãäåçèéêëìíîïñòóôõöùúûüýÿÀÁÂÃÄÅÇÈÉÊËÌÍÎÏÑÒÓÔÕÖÙÚÛÜÝ';
  const String to = 'aaaaaaceeeeiiiinooooouuuuyyAAAAAACEEEEIIIINOOOOOUUUUY';
  final StringBuffer b = StringBuffer();
  for (int i = 0; i < s.length; i++) {
    final String c = s[i];
    final int k = from.indexOf(c);
    b.write(k >= 0 ? to[k] : c);
  }
  return b
      .toString()
      .toUpperCase()
      .replaceAll(RegExp(r'[^A-Z0-9]+'), ' ')
      .trim();
}

class RegisteredSource {
  const RegisteredSource({
    required this.service,
    required this.role,
    required this.rank,
    required this.level,
    required this.network,
  });

  final ScheduleSource service;
  final String role;
  final int rank;
  final String level;
  final String? network;

  FeedProvenance get provenance => service.provenance;
}

class ResolvedSource {
  const ResolvedSource(this.source, this.validityStatus);

  final RegisteredSource source;
  final String validityStatus;

  bool get usable => validityStatus == Validity.current;
}

class ProviderReference {
  const ProviderReference(this.serviceDate, this.time);

  final String serviceDate;
  final String time;

  String get isoDate => toIsoDate(serviceDate);
}

class RouteMatch {
  const RouteMatch(this.route, this.source);

  final GtfsRoute route;
  final RegisteredSource source;
}

class StopMatches {
  const StopMatches(this.exact, this.partial);

  final List<GtfsStop> exact;
  final List<GtfsStop> partial;
}

class TransitDataProvider {
  TransitDataProvider({DateTime Function()? now})
      : now = now ?? DateTime.now;

  final DateTime Function() now;
  final List<RegisteredSource> sources = <RegisteredSource>[];

  /// Enregistre une source avec un rôle. Règles strictes :
  ///  - un feed déclaré HISTORICAL ne peut recevoir qu'un rôle historicalReference ;
  ///  - un rôle « current » exige un statut déclaré CURRENT ;
  ///  - currentOfficial exige une source institutionnelle (CETUD / autorité).
  RegisteredSource registerSource(ScheduleSource service, {required String role}) {
    if (!ProviderRoles.all.contains(role)) {
      throw ArgumentError('TransitDataProvider : rôle inconnu « $role »');
    }
    final FeedProvenance prov = service.provenance;
    final String label =
        '${prov.source}${prov.network == null ? '' : ' (${prov.network})'}';
    if (role != ProviderRoles.historicalReference) {
      if (prov.declaredStatus == Validity.historical) {
        throw StateError(
            'TransitDataProvider : $label est déclarée HISTORICAL et ne peut pas être enregistrée comme $role');
      }
      if (prov.declaredStatus != Validity.current) {
        throw StateError(
            'TransitDataProvider : $label n\'a pas de statut CURRENT prouvé (statut ${prov.declaredStatus}) — rôle $role refusé');
      }
      final String? requiredType = ProviderRoles.requiredSourceType[role];
      if (requiredType != null && prov.sourceType != requiredType) {
        throw StateError(
            'TransitDataProvider : $label (${prov.sourceTypeLabel}) ne peut pas tenir le rôle $role ($requiredType requis)');
      }
      if (role == ProviderRoles.currentFrequency && service.kind != 'frequency') {
        throw StateError(
            'TransitDataProvider : le rôle currentFrequency est réservé aux sources de fréquence');
      }
      if (role != ProviderRoles.currentFrequency && service.kind == 'frequency') {
        throw StateError(
            'TransitDataProvider : une source de fréquence ne peut pas tenir le rôle $role');
      }
    }
    final RegisteredSource entry = RegisteredSource(
      service: service,
      role: role,
      rank: _rankFor(role, prov),
      level: ProviderRoles.level[role]!,
      network: service.network ?? prov.network,
    );
    sources.add(entry);
    sources.sort((RegisteredSource a, RegisteredSource b) => a.rank.compareTo(b.rank));
    return entry;
  }

  int _rankFor(String role, FeedProvenance prov) {
    switch (role) {
      case ProviderRoles.currentOfficial:
        return normalizeText(prov.authority ?? prov.source) == 'CETUD' ? 0 : 1;
      case ProviderRoles.currentApplication:
        return 2;
      case ProviderRoles.currentOpenData:
        return 3;
      case ProviderRoles.currentFrequency:
        return 4;
      default:
        return 9;
    }
  }

  /// Date/heure de référence (horloge Dakar, UTC+0).
  ProviderReference reference({String? date, String? time}) {
    final DakarClock clock = dakarClock(now());
    final String serviceDate =
        date == null ? clock.serviceDate : normalizeServiceDate(date);
    final String t = time ?? clock.time;
    final int? seconds = parseGtfsTime(t);
    if (seconds == null) {
      throw FormatException('TransitDataProvider : heure invalide « $t »');
    }
    return ProviderReference(serviceDate, formatGtfsTime(seconds));
  }

  /// Sources « current » avec leur validité à une date.
  List<ResolvedSource> resolveCurrentSources(String asOf) {
    final String day = normalizeServiceDate(asOf);
    return sources
        .where((RegisteredSource s) => ProviderRoles.current.contains(s.role))
        .map((RegisteredSource s) =>
            ResolvedSource(s, s.provenance.validityStatusOn(day)))
        .toList();
  }

  List<RegisteredSource> historicalSources() => sources
      .where((RegisteredSource s) => s.role == ProviderRoles.historicalReference)
      .toList();

  /// Source gagnante pour un réseau à une date (null → UNKNOWN).
  RegisteredSource? winningSource(String? network, String asOf) {
    final String? net = network?.toUpperCase();
    for (final ResolvedSource r in resolveCurrentSources(asOf)) {
      if (!r.usable || r.source.service.kind == 'frequency') continue;
      final String? sn = r.source.network?.toUpperCase();
      if (net == null || sn == null || sn == net) return r.source;
    }
    return null;
  }

  DepartureResult _unknown(
    String routeId,
    String? stopId,
    ProviderReference ref,
    String reason,
  ) =>
      DepartureResult(
        routeId: routeId,
        stopId: stopId,
        date: ref.isoDate,
        currentTime: ref.time,
        status: ScheduleStatus.unknown,
        reason: reason,
        source: null,
        sourceType: null,
        feedVersion: null,
        validity: null,
        provenanceLevel: ProvenanceLevels.unknown,
        departures: const <Departure>[],
        historicalReferenceAvailable:
            historicalSources().any((RegisteredSource s) => s.service.knowsRoute(routeId)),
      );

  DepartureResult _decorate(DepartureResult r, RegisteredSource s) {
    final bool current =
        r.status == ScheduleStatus.scheduled || r.status == ScheduleStatus.estimated;
    return r.copyWith(
      isCurrent: current,
      provenanceLevel: r.status == ScheduleStatus.unknown ? ProvenanceLevels.unknown : s.level,
      role: s.role,
      network: s.network,
      departures: r.departures.map((Departure d) => d.withLevel(s.level)).toList(),
    );
  }

  /// Prochains départs « actuels » : uniquement sources current utilisables.
  DepartureResult getDepartures(
    String routeId,
    String? stopId, {
    String? date,
    String? time,
    int? limit = 5,
  }) {
    final ProviderReference ref = reference(date: date, time: time);
    DepartureResult? fallback;
    for (final ResolvedSource r in resolveCurrentSources(ref.serviceDate)) {
      if (!r.usable || !r.source.service.knowsRoute(routeId)) continue;
      final DepartureResult res = r.source.service.getDeparturesAtStop(
        routeId,
        stopId,
        ref.serviceDate,
        ref.time,
        limit: limit,
        asOf: ref.serviceDate,
      );
      if (res.status == ScheduleStatus.scheduled || res.status == ScheduleStatus.estimated) {
        return _decorate(res, r.source);
      }
      fallback ??= _decorate(res, r.source);
    }
    return fallback ?? _unknown(routeId, stopId, ref, ScheduleReasons.noCurrentSource);
  }

  DepartureResult getDeparturesNow(String routeId, String? stopId, {int? limit = 5}) =>
      getDepartures(routeId, stopId, limit: limit);

  /// Consultation explicite de la référence historique (jamais « actuel »).
  DepartureResult getHistoricalDepartures(
    String routeId,
    String? stopId, {
    String? date,
    String? time,
    int? limit = 5,
  }) {
    final ProviderReference ref = reference(date: date, time: time);
    for (final RegisteredSource s in historicalSources()) {
      if (!s.service.knowsRoute(routeId)) continue;
      final DepartureResult res = s.service.getDeparturesAtStop(
        routeId,
        stopId,
        ref.serviceDate,
        ref.time,
        limit: limit,
        asOf: ref.serviceDate,
      );
      return res.copyWith(
        isCurrent: false,
        provenanceLevel: ProvenanceLevels.historicalReference,
        role: ProviderRoles.historicalReference,
      );
    }
    return _unknown(routeId, stopId, ref, ScheduleReasons.routeUnknown);
  }

  /// Lignes dont route_short_name (champ officiel) vaut exactement le numéro.
  List<RouteMatch> routesForLine(
    String? network,
    String lineNumber, {
    String? asOf,
    bool includeHistorical = false,
  }) {
    final String wanted = normalizeText(lineNumber);
    if (wanted.isEmpty) return <RouteMatch>[];
    final String? net = network?.toUpperCase();
    final String day = asOf == null
        ? dakarClock(now()).serviceDate
        : normalizeServiceDate(asOf);
    final List<RegisteredSource> pool = includeHistorical
        ? historicalSources()
        : resolveCurrentSources(day)
            .where((ResolvedSource r) => r.usable)
            .map((ResolvedSource r) => r.source)
            .toList();
    final List<RouteMatch> out = <RouteMatch>[];
    for (final RegisteredSource s in pool) {
      if (s.service.kind == 'frequency') continue;
      final String? sn = s.network?.toUpperCase();
      if (net != null && sn != null && sn != net) continue;
      for (final GtfsRoute r in s.service.routes()) {
        if (normalizeText(r.routeShortName) == wanted) out.add(RouteMatch(r, s));
      }
    }
    return out;
  }

  /// Arrêts dont le nom correspond (exact normalisé, sinon inclusion).
  static StopMatches matchStops(List<GtfsStop> stops, String name) {
    final String wanted = normalizeText(name);
    if (wanted.isEmpty) return const StopMatches(<GtfsStop>[], <GtfsStop>[]);
    final List<GtfsStop> exact =
        stops.where((GtfsStop s) => normalizeText(s.stopName) == wanted).toList();
    if (exact.isNotEmpty) return StopMatches(exact, <GtfsStop>[]);
    final List<GtfsStop> partial =
        stops.where((GtfsStop s) => normalizeText(s.stopName).contains(wanted)).toList();
    return StopMatches(<GtfsStop>[], partial);
  }
}

File: lib/external-gtfs/transit-data-provider.js
'use strict';
/**
 * TransitDataProvider : couche unique de résolution des horaires.
 *
 * Ordre de priorité (spécification LOT 18 BIS §8) :
 *   1. CETUD CURRENT (currentOfficial, autorité CETUD)
 *   2. autre source officielle actuelle (currentOfficial, autre autorité)
 *   3. PassBi CURRENT si réellement prouvé (currentApplication)
 *   4. open data actuelle (currentOpenData)
 *   5. estimation par fréquence documentée (currentFrequency → ESTIMATED)
 *   6. UNKNOWN
 * Les sources HISTORICAL (PassBi) ne participent JAMAIS à une réponse
 * « maintenant » : elles ne sont interrogeables que via getHistoricalDepartures.
 */
const { normalizeServiceDate, toIsoDate, dakarClock, formatTime, parseTime } = require('./gtfs-time');
const { SOURCE_TYPES, VALIDITY } = require('./provenance');
const { STATUS, REASONS } = require('./gtfs-schedule-service');

const ROLES = Object.freeze({
  CURRENT_OFFICIAL: 'currentOfficial',
  CURRENT_APPLICATION: 'currentApplication',
  CURRENT_OPEN_DATA: 'currentOpenData',
  CURRENT_FREQUENCY: 'currentFrequency',
  HISTORICAL_REFERENCE: 'historicalReference',
});
const ROLE_LIST = Object.freeze(Object.values(ROLES));
const CURRENT_ROLES = Object.freeze([ROLES.CURRENT_OFFICIAL, ROLES.CURRENT_APPLICATION, ROLES.CURRENT_OPEN_DATA, ROLES.CURRENT_FREQUENCY]);

const ROLE_SOURCE_TYPE = Object.freeze({
  [ROLES.CURRENT_OFFICIAL]: SOURCE_TYPES.SOURCE_INSTITUTIONAL,
  [ROLES.CURRENT_APPLICATION]: SOURCE_TYPES.SOURCE_APPLICATION,
  [ROLES.CURRENT_OPEN_DATA]: SOURCE_TYPES.OPEN_DATA,
});

const ROLE_LEVEL = Object.freeze({
  [ROLES.CURRENT_OFFICIAL]: 'OFFICIAL_STATIC_CURRENT',
  [ROLES.CURRENT_APPLICATION]: 'SOURCE_APPLICATION_CURRENT',
  [ROLES.CURRENT_OPEN_DATA]: 'OPEN_DATA_CURRENT',
  [ROLES.CURRENT_FREQUENCY]: 'ESTIMATED',
  [ROLES.HISTORICAL_REFERENCE]: 'HISTORICAL_REFERENCE',
});

const normalizeText = (s) => String(s || '')
  .normalize('NFD').replace(/[\u0300-\u036f]/g, '')
  .toUpperCase().replace(/[^A-Z0-9]+/g, ' ').trim();

class TransitDataProvider {
  /** @param {{now?: () => Date}} [options] horloge injectable (tests) */
  constructor(options = {}) {
    this.now = typeof options.now === 'function' ? options.now : () => new Date();
    this.sources = [];
  }

  /**
   * Enregistre une source horaire avec un rôle. Règles strictes :
   *  - un feed déclaré HISTORICAL ne peut recevoir qu'un rôle historicalReference ;
   *  - un rôle « current » exige un statut déclaré CURRENT ;
   *  - currentOfficial exige une source institutionnelle (CETUD / autorité).
   */
  registerSource(service, { role, realtime = false } = {}) {
    if (!service || typeof service.getDeparturesAtStop !== 'function' || !service.provenance) {
      throw new Error('TransitDataProvider : source invalide (getDeparturesAtStop / provenance requis)');
    }
    if (!ROLE_LIST.includes(role)) throw new Error(`TransitDataProvider : rôle inconnu « ${role} »`);
    const prov = service.provenance;
    const label = `${prov.source}${prov.network ? ` (${prov.network})` : ''}`;
    if (role !== ROLES.HISTORICAL_REFERENCE) {
      if (prov.declaredStatus === VALIDITY.HISTORICAL) {
        throw new Error(`TransitDataProvider : ${label} est déclarée HISTORICAL et ne peut pas être enregistrée comme ${role}`);
      }
      if (prov.declaredStatus !== VALIDITY.CURRENT) {
        throw new Error(`TransitDataProvider : ${label} n'a pas de statut CURRENT prouvé (statut ${prov.declaredStatus}) — rôle ${role} refusé`);
      }
      if (ROLE_SOURCE_TYPE[role] && prov.sourceType !== ROLE_SOURCE_TYPE[role]) {
        throw new Error(`TransitDataProvider : ${label} (${prov.sourceTypeLabel}) ne peut pas tenir le rôle ${role} (${ROLE_SOURCE_TYPE[role]} requis)`);
      }
      if (role === ROLES.CURRENT_FREQUENCY && service.kind !== 'frequency') {
        throw new Error(`TransitDataProvider : le rôle currentFrequency est réservé aux sources de fréquence`);
      }
      if (role !== ROLES.CURRENT_FREQUENCY && service.kind === 'frequency') {
        throw new Error(`TransitDataProvider : une source de fréquence ne peut pas tenir le rôle ${role}`);
      }
    }
    const entry = {
      service,
      role,
      realtime: Boolean(realtime),
      provenance: prov,
      network: service.network || prov.network || null,
      rank: this.rankFor(role, prov),
      level: role === ROLES.CURRENT_OFFICIAL && realtime ? 'OFFICIAL_CURRENT' : ROLE_LEVEL[role],
    };
    this.sources.push(entry);
    this.sources.sort((a, b) => a.rank - b.rank);
    return entry;
  }

  rankFor(role, prov) {
    switch (role) {
      case ROLES.CURRENT_OFFICIAL: return normalizeText(prov.authority || prov.source) === 'CETUD' ? 0 : 1;
      case ROLES.CURRENT_APPLICATION: return 2;
      case ROLES.CURRENT_OPEN_DATA: return 3;
      case ROLES.CURRENT_FREQUENCY: return 4;
      default: return 9;
    }
  }

  /** Date/heure de référence (horloge Dakar, UTC+0). */
  reference(options = {}) {
    const clock = dakarClock(this.now());
    const serviceDate = options.date ? normalizeServiceDate(options.date) : clock.serviceDate;
    const time = options.time || clock.time;
    if (parseTime(time) === null) throw new Error(`TransitDataProvider : heure invalide « ${time} »`);
    return { serviceDate, isoDate: toIsoDate(serviceDate), time: formatTime(parseTime(time)) };
  }

  /** Sources « current » utilisables à une date (validité vérifiée à chaque appel). */
  resolveCurrentSources(asOf) {
    const day = normalizeServiceDate(asOf);
    return this.sources
      .filter((s) => CURRENT_ROLES.includes(s.role))
      .map((s) => {
        const status = s.provenance.validityStatusOn(day);
        return { ...s, validityStatus: status, usable: status === VALIDITY.CURRENT };
      });
  }

  historicalSources() { return this.sources.filter((s) => s.role === ROLES.HISTORICAL_REFERENCE); }

  /** Source gagnante pour un réseau à une date (ou null → UNKNOWN). */
  winningSource(network, asOf) {
    const net = network ? String(network).toUpperCase() : null;
    return this.resolveCurrentSources(asOf).find((s) => s.usable && s.service.kind !== 'frequency' && (!net || !s.network || String(s.network).toUpperCase() === net)) || null;
  }

  /**
   * Prochains départs « actuels » : uniquement sources current utilisables ;
   * SCHEDULED depuis stop_times, sinon ESTIMATED (fréquence documentée), sinon UNKNOWN.
   */
  getDepartures(routeId, stopId, options = {}) {
    const ref = this.reference(options);
    const limit = options.limit === undefined ? 5 : options.limit;
    const current = this.resolveCurrentSources(ref.serviceDate);
    const consulted = [];
    let fallback = null;
    for (const s of current) {
      const knows = s.service.knowsRoute(routeId);
      consulted.push({ source: s.provenance.source, network: s.network, role: s.role, validityStatus: s.validityStatus, usable: s.usable, knowsRoute: knows });
      if (!s.usable || !knows) continue;
      const result = s.service.getDeparturesAtStop(routeId, stopId, ref.serviceDate, ref.time, { limit, asOf: ref.serviceDate });
      if (result.status === STATUS.SCHEDULED || result.status === STATUS.ESTIMATED) {
        return this.decorate(result, s, ref, consulted);
      }
      if (!fallback) fallback = this.decorate(result, s, ref, consulted);
    }
    if (fallback) return { ...fallback, consulted };
    return this.unknown(routeId, stopId, ref, consulted, current.length ? REASONS.NO_CURRENT_SOURCE : REASONS.NO_CURRENT_SOURCE);
  }

  getDeparturesNow(routeId, stopId, options = {}) {
    return this.getDepartures(routeId, stopId, { ...options, date: undefined, time: undefined });
  }

  /** Consultation explicite de la référence historique (jamais « actuel »). */
  getHistoricalDepartures(routeId, stopId, options = {}) {
    const ref = this.reference(options);
    const limit = options.limit === undefined ? 5 : options.limit;
    const sources = this.historicalSources().filter((s) => !options.sourceId || s.provenance.sourceId === options.sourceId);
    for (const s of sources) {
      if (!s.service.knowsRoute(routeId)) continue;
      const result = s.service.getDeparturesAtStop(routeId, stopId, ref.serviceDate, ref.time, { limit, asOf: ref.serviceDate });
      return {
        ...result,
        isCurrent: false,
        provenanceLevel: 'HISTORICAL_REFERENCE',
        role: ROLES.HISTORICAL_REFERENCE,
        note: 'Référence historique : ne décrit pas le service actuel.',
      };
    }
    return this.unknown(routeId, stopId, ref, [], REASONS.ROUTE_UNKNOWN);
  }

  decorate(result, entry, ref, consulted) {
    return {
      ...result,
      isCurrent: result.status === STATUS.SCHEDULED || result.status === STATUS.ESTIMATED,
      provenanceLevel: result.status === STATUS.UNKNOWN ? 'UNKNOWN' : entry.level,
      role: entry.role,
      network: entry.network,
      consulted,
      departures: (result.departures || []).map((d) => ({ ...d, provenanceLevel: entry.level })),
    };
  }

  unknown(routeId, stopId, ref, consulted, reason) {
    return {
      routeId,
      stopId,
      date: ref.isoDate,
      currentTime: ref.time,
      status: STATUS.UNKNOWN,
      reason,
      source: null,
      sourceType: null,
      feedVersion: null,
      validity: null,
      provenanceLevel: 'UNKNOWN',
      isCurrent: false,
      role: null,
      consulted,
      historicalReferenceAvailable: this.historicalSources().some((s) => s.service.knowsRoute(routeId)),
      departures: [],
    };
  }

  /** Lignes d'un réseau dont route_short_name (champ officiel) vaut exactement le numéro demandé. */
  routesForLine(network, lineNumber, { asOf, includeHistorical = false } = {}) {
    const wanted = normalizeText(lineNumber);
    if (!wanted) return [];
    const net = network ? String(network).toUpperCase() : null;
    const day = asOf ? normalizeServiceDate(asOf) : dakarClock(this.now()).serviceDate;
    const pool = includeHistorical ? this.historicalSources().map((s) => ({ ...s, usable: true })) : this.resolveCurrentSources(day).filter((s) => s.usable);
    const out = [];
    for (const s of pool) {
      if (s.service.kind === 'frequency') continue; // les fréquences ne sont pas des lignes GTFS (voir frequencyEstimate)
      if (net && s.network && String(s.network).toUpperCase() !== net) continue;
      for (const r of s.service.routes()) {
        if (normalizeText(r.routeShortName) === wanted) out.push({ route: r, source: s });
      }
    }
    return out;
  }

  /** Arrêts d'une source dont le nom correspond (exact normalisé, puis inclusion unique). */
  static matchStops(stops, name) {
    const wanted = normalizeText(name);
    if (!wanted) return { exact: [], partial: [] };
    const exact = stops.filter((s) => normalizeText(s.stopName) === wanted);
    if (exact.length) return { exact, partial: [] };
    const partial = stops.filter((s) => normalizeText(s.stopName).includes(wanted));
    return { exact: [], partial };
  }

  describe(asOf) {
    const day = asOf ? normalizeServiceDate(asOf) : dakarClock(this.now()).serviceDate;
    return {
      asOf: toIsoDate(day),
      sources: this.sources.map((s) => ({
        source: s.provenance.source,
        network: s.network,
        role: s.role,
        rank: s.rank,
        level: s.level,
        sourceType: s.provenance.sourceTypeLabel,
        declaredStatus: s.provenance.declaredStatus,
        validity: s.provenance.validityOn(day),
        usableNow: CURRENT_ROLES.includes(s.role) && s.provenance.isCurrentOn(day),
      })),
    };
  }
}

module.exports = { TransitDataProvider, ROLES, ROLE_LEVEL, CURRENT_ROLES, normalizeText };
